pallets/file-system/src/lib.rs

```diff
@@ -284,38 +302,60 @@ pub mod pallet {
+                    db_weight.reads_writes(1, T::MaxExpiredStorageRequests::get().into());
+
+                // Check if there is enough weight left to process the cleanup for one more block
+                if required_weight.all_gte(remaining_weight.saturating_sub(total_used_weight)) {
+                    break;
+                }
+
+                let mut used_weight = db_weight.reads(1);
+                let expired_requests = StorageRequestExpirations::<T>::take(&block_to_clean);
+
+                // Remove expired storage requests for the block
+                for location in expired_requests {
+                    StorageRequests::<T>::remove(&location);
+                    used_weight += db_weight.writes(1);
+                    Self::deposit_event(Event::StorageRequestExpired { location });
+                }
+
+                // Accumulate the weight used for cleanup operations
+                total_used_weight += used_weight;
+                // Increment the block to clean up for the next iteration
+                block_to_clean = block_to_clean.saturating_add(1u32.into());
             }
 
-            let mut used_weight = db_weight.reads(1);
-
-            // TODO: test the behaviour of an on_idle execution when the previous one didn't have enough remaining weight
-            // and the current block doesn't match the previous one. (do we need to add extra logic here to take into account those scenarios?)
-            let mut expired_requests = StorageRequestExpirations::<T>::take(&block);
-
-            // Remove expired storage requests
-            for location in expired_requests.drain(..) {
-                // TODO: should probably add some fields to the `StorageRequestExpired` to facilitate SPs filtering the events
-                // that are relevant to them (e.g. include the SPs that have volunteered to store the file)
-                let _request = StorageRequests::<T>::take(&location);
-
-                StorageRequests::<T>::remove(&location);
-                used_weight += db_weight.writes(1);
-
-                Self::deposit_event(Event::StorageRequestExpired { location });
+            // Update `NextStartingBlockToCleanUp`
+            match NextStartingBlockToCleanUp::<T>::get() {
+                // `NextStartingBlockToCleanUp` is always updated to start from the block we reached in the current `on_idle` call.
+                Some(start_block) if block_to_clean > start_block => {
+                    NextStartingBlockToCleanUp::<T>::put(block_to_clean);
+                }
+                // `NextStartingBlockToCleanUp` is initialized here for the first time if we have not cleaned up the current block.
+                None if block_to_clean == current_block => {
+                    NextStartingBlockToCleanUp::<T>::put(block_to_clean);
+                }
+                // In all other cases, this means we processed all the blocks up to the current block
+                _ => {}
```

Comment:
This is prone to confusion. I believe `NextStartingBlockToCleanUp::<T>::get()` should actually be `ValueQuery` and return `0` if it has never been accessed. That way we make sure that the cleanup always starts checking storage requests to delete right from the first block.

Reply:
Ya I agree with this.